### PR TITLE
[Bug fix] - pass sub_core_grids in sampling masking (fixes  #48308 )

### LINKED ISSUE #48308: [models]: sampling broken with sub_core_grids
### Component / Area

models

### Issue Type (optional)

Runtime Crash

### Observed

After #47021 , sampling crashes with an error that, from experience, means sub_core_grids is not set correctly. This is exposed eg. in vllm nightlyhttps://github.com/tenstorrent/tt-metal/actions/runs/28209887137/job/83570685619#step:19:480

### Expected

correctly pass sub_core_grids

### 1. Steps (exact commands)

run llama3_70b_galaxy. all_model_tests should also catch this - I see the first run failed with machine issues, and it wasn't re-ran correctly later.

### 2. Input data / link or description

-

### 3. Frequency

always

### 1. Software Versions

-

### 2. Hardware Details

-

### Is this a regression?

Yes

### Regression Details

#47021 

### Logs & Diagnostics

_No response_

### Priority

P1

### Impact

llama3_70b_galaxy broken

### PR DESCRIPTION
**Pull request by tchedaTT**
### PR Category
Bug fix
### Summary
Fixes #48308 

### Notes for reviewers
-

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:tcheda/fix-48308-sampling-sub-core-grids)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:tcheda/fix-48308-sampling-sub-core-grids)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:tcheda/fix-48308-sampling-sub-core-grids)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:tcheda/fix-48308-sampling-sub-core-grids)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:tcheda/fix-48308-sampling-sub-core-grids)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:tcheda/fix-48308-sampling-sub-core-grids)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:tcheda/fix-48308-sampling-sub-core-grids)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:tcheda/fix-48308-sampling-sub-core-grids)

### CODE REVIEW

**Review by sraizada-tt — APPROVED** [review]

**Comment by tchedaTT** [comment]
CI in progress
T3K unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/28372511904
vLLM nightly:  https://github.com/tenstorrent/tt-metal/actions/runs/28372795885
All model tests slice: https://github.com/tenstorrent/tt-metal/actions/runs/28372919725

File: models/common/sampling/tt_sampling.py [hunk 1]
```diff
@@ -375,7 +375,12 @@ def _mask_invalid_vocab_logits(self, logits):
             return self._mask_invalid_vocab_tail_logits(logits)
         if self.tt_invalid_vocab_mask is None:
             return logits
-        return ttnn.add(logits, self.tt_invalid_vocab_mask, memory_config=logits.memory_config())
+        return ttnn.add(
+            logits,
+            self.tt_invalid_vocab_mask,
+            memory_config=logits.memory_config(),
+            sub_core_grids=self.sub_core_grids,
+        )
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply models/common/sampling/tt_sampling.py:383]: llama-galaxy unit tests are commented out in models_unit_tests, so I couldnt even run a test if I added it.
This path will be hit in common integration tests, running those

**Review by mtairum — APPROVED** [review]

**Comment by mtairum** [comment]
Approved to unblock. Make sure we run the CI tests to confirm the changes.

**Comment by mtairum** [comment]
Claude  review:

 directionally this is right (the masking ops should match the `ttnn.slice` calls, which already pass `sub_core_grids`), and covering both `TTSampling` and `Sampling1D` plus the fallback paths is nice and thorough. Two things I'd like to resolve before approving:

**1. The `ttnn.concat` is left without `sub_core_grids`.** The original crash traceback points exactly at the concat in `_mask_invalid_vocab_tail_logits` (tt_sampling.py:406, `ttnn::prim::concat → TT_FATAL num_intersections == num_cores`), and for Llama-3.3-70B on Galaxy the `valid_width > 0` path does reach that concat. This PR fixes the upstream `ttnn.add` placement but leaves the concat as-is, so it's relying on the concat inheriting a compatible grid from its now-correctly-placed inputs. For interleaved inputs without `sub_core_grids`, concat goes through the `massaged_concat` path, which builds its program on a device-chosen grid rather than the inputs' sub-core grid — so there's a real chance the same TT_FATAL recurs under an active sub-device manager. `ttnn.concat` accepts `sub_core_grids`; could we pass it there too (belt-and-suspenders) so the actual crashing op is explicitly constrained?

**2. The new unit test doesn't reproduce this regression.** `test_sampling1d_padded_tail_with_sub_core_grids_runs` runs on a `(1,8)` mesh and only passes `sub_core_grids` to the sampler — it never sets up an **active sub-device manager**, which is what actually makes this error fire ("kernel group cores do not match *sub device* cores"). Without an active sub-device, ops can legally use the full grid, so this test would pass on `main` too. It also only asserts `tokens.numel() == B` rather than `tokens < valid_vocab_size`. Could we either exercise it under an active sub-device manager, or at minimum assert the padded tokens are masked out?

Functionally I expect this to be fine for the crash (which is the scope here — the separate determinism failures in the vLLM sampling job are a different issue). I'd just like the Galaxy vLLM nightly currently running on this branch to confirm the Llama-3.3-70B sampling job gets past `_mask_invalid_vocab_tail_logits` before we merge, since the concat is the one unproven link.

**Comment by tchedaTT** [comment]
Can't run unit test because it's blocked by a preexisting broken mlp test and a fail fast policy.
All-model-tests failed with a machine issue, but main can't run it anyway https://github.com/tenstorrent/tt-metal/actions/runs/28346542165/job/83971896402#step:6:1022
Nightly failed twice on machine issue before fianlly actually getting to the test.
Concat needs the additional sub_core_grids, test in flight at https://github.com/tenstorrent/tt-metal/actions/runs/28394289185/job/84128931914

**Comment by tchedaTT** [comment]
Tests still not passing cleanly - showing same pattern of hangs as since April (probably ND, since April 1 the job passed 7 out of 89 nightlies). Investigating whether I can find a mitigation - if not I'm still probably going to merge, as it's less broken than before.

**Comment by tchedaTT** [comment]
Looks like it's a real ND hang. Happens in the same spot as before the sub_core_grids regression was introduced. 
5 days ago https://github.com/tenstorrent/tt-metal/actions/runs/28139255072/job/83335013970#step:19:6945
Now https://github.com/tenstorrent/tt-metal/actions/runs/28411372753/job/84188508805#step:19:2945
I'm in favor of merging, I'd like a second opinion @mtairum @sraizada-tt

**Review by yieldthought — APPROVED** [review]

**Comment by sraizada-tt** [comment]
can merge
